radix-common/src/state/state_updates.rs

```diff
@@ -0,0 +1,282 @@
+            new_substate_values: index_map_new(),
+        });
+    }
+
+    /// Applies the given updates on top of the current updates to the partition.
+    pub fn update_substates(
+        &mut self,
+        updates: impl IntoIterator<Item = (SubstateKey, DatabaseUpdate)>,
+    ) {
+        match self {
+            PartitionStateUpdates::Delta { by_substate } => by_substate.extend(updates),
+            PartitionStateUpdates::Batch(BatchPartitionStateUpdate::Reset {
+                new_substate_values,
+            }) => {
+                for (substate_key, database_update) in updates {
+                    match database_update {
+                        DatabaseUpdate::Set(new_value) => {
+                            new_substate_values.insert(substate_key, new_value);
+                        }
+                        DatabaseUpdate::Delete => {
+                            let existed = new_substate_values.swap_remove(&substate_key).is_some();
+                            if !existed {
+                                panic!("inconsistent update: delete of substate {:?} not existing in reset partition", substate_key);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    pub fn without_empty_entries(&self) -> Option<Self> {
+        match self {
+            PartitionStateUpdates::Delta { by_substate } => {
+                if by_substate.len() > 0 {
+                    Some(PartitionStateUpdates::Delta {
+                        by_substate: by_substate.clone(),
+                    })
+                } else {
+                    None
+                }
+            }
+            PartitionStateUpdates::Batch(x) => {
+                // We shouldn't filter out batch updates like resets, even if they set nothing new
+                Some(PartitionStateUpdates::Batch(x.clone()))
+            }
+        }
+    }
+}
+
+/// A description of a batch update affecting an entire Partition.
+#[derive(Debug, Clone, PartialEq, Eq, Sbor)]
+pub enum BatchPartitionStateUpdate {
+    /// A reset, dropping all Substates of a partition and replacing them with a new set.
+    /// Contains indexed new Substate values, captured in the order of creation of a Substate.
+    Reset {
+        new_substate_values: IndexMap<SubstateKey, DbSubstateValue>,
+    },
+}
+
+/// An update of a single substate's value.
+#[derive(Debug, Clone, Hash, PartialEq, Eq, Sbor, PartialOrd, Ord)]
+pub enum DatabaseUpdate {
+    Set(DbSubstateValue),
+    Delete,
+}
+
+impl DatabaseUpdate {
+    pub fn as_ref(&self) -> DatabaseUpdateRef<'_> {
+        match self {
+            DatabaseUpdate::Set(update) => DatabaseUpdateRef::Set(update),
+            DatabaseUpdate::Delete => DatabaseUpdateRef::Delete,
+        }
+    }
+}
+
+/// A 1:1 counterpart of [`DatabaseUpdate`], but operating on references.
+pub enum DatabaseUpdateRef<'v> {
```

Comment:
nice; if we are touching this, should we go for:
```
pub enum DatabaseUpdateWithValueOrRef<V> {
...
}

pub type DatabaseUpdate = DatabaseUpdateWithValueOrRef<DbSubstateValue>;

pub type DatabaseUpdateRef<'v> = DatabaseUpdateWithValueOrRef<&'v [u8]>;
```

(please disregard if it does not deduplicate any "common" handling code and just makes things harder ;p)

Reply:
Once we have https://github.com/radixdlt/radixdlt-scrypto/pull/1860 I think it's natural to formulate it as:
```
pub enum DatabaseUpdate<'v> {
    Set(RawScryptoPayload<'v>),
    Delete,
}
pub type OwnedDatabaseUpdate = DatabaseUpdate<'static>;
```
(Happy to debate naming) So perhaps we can wait for that?

Reply:
Yes, We Can! ®️